dali/operators/segmentation/biased_crop_center.cc

```diff
@@ -0,0 +1,180 @@
+pixels in the input. If the selected crop center results in an out of bounds cropping window,
+the center is shifted as necessary so that the window remains within bounds.
+
+If left unspecified, the shape is not taken into account when selecting the center,
+effectively resulting in any position in the input being a valid center.
+)code",
+      nullptr, true)
+    .NumInput(1)
+    .NumOutput(1);
+
+class BiasedCropCenterCPU : public Operator<CPUBackend> {
+ public:
+  explicit BiasedCropCenterCPU(const OpSpec &spec);
+  bool CanInferOutputs() const override { return true; }
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<CPUBackend> &ws) override;
+  void RunImpl(workspace_t<CPUBackend> &ws) override;
+
+ private:
+  template <typename T>
+  void RunImplTyped(workspace_t<CPUBackend> &ws);
+
+  std::mt19937_64 rng_;
+
+  std::vector<int> foreground_;
+  TensorListShape<> crop_shape_;
+
+  bool has_crop_shape_ = false;
+
+  USE_OPERATOR_MEMBERS();
+};
+
+BiasedCropCenterCPU::BiasedCropCenterCPU(const OpSpec &spec)
+    : Operator<CPUBackend>(spec), rng_(spec.GetArgument<int64_t>("seed")) {
+}
+
+bool BiasedCropCenterCPU::SetupImpl(std::vector<OutputDesc> &output_desc,
+                                    const workspace_t<CPUBackend> &ws) {
+  const auto &in_masks = ws.template InputRef<CPUBackend>(0);
+  int nsamples = in_masks.size();
+  auto in_masks_shape = in_masks.shape();
+  int ndim = in_masks_shape.sample_dim();
+  output_desc.resize(1);
+  output_desc[0].shape = uniform_list_shape(nsamples, {ndim});
+  output_desc[0].type = TypeTable::GetTypeInfo(DALI_INT64);
+
+  foreground_.clear();
+  if (spec_.HasTensorArgument("foreground")) {
+    foreground_.resize(nsamples);
+    const auto &foreground_arg_in = ws.ArgumentInput("foreground");
+    for (int i = 0; i < nsamples; i++) {
+      foreground_[i] = *foreground_arg_in[i].data<int>();
+    }
+  } else {
+    foreground_.push_back(spec_.GetArgument<int>("foreground"));
+  }
+
+  has_crop_shape_ = spec_.ArgumentDefined("shape");
+  if (has_crop_shape_) {
+    GetShapeArgument(crop_shape_, spec_, "shape", ws, ndim, nsamples);
+  }
+  return true;
+}
+
+template <typename T>
+void BiasedCropCenterCPU::RunImplTyped(workspace_t<CPUBackend> &ws) {
+  const auto &in_masks = ws.template InputRef<CPUBackend>(0);
+  auto &out_center = ws.template OutputRef<CPUBackend>(0);
+  int nsamples = in_masks.size();
+  auto in_masks_shape = in_masks.shape();
+  int ndim = in_masks_shape.sample_dim();
+  auto masks_view = view<const T>(in_masks);
+  auto center_view = view<int64_t>(out_center);
+  auto& thread_pool = ws.GetThreadPool();
+
+  for (int sample_idx = 0; sample_idx < nsamples; sample_idx++) {
+    thread_pool.AddWork(
+      [&, sample_idx](int thread_id) {
+        auto mask = masks_view[sample_idx];
+        auto center = center_view[sample_idx];
+        SearchableRLEMask rle_mask(mask);
```

Comment:
Can you tell how fast/slow it is for the typical inputs?
Do we assume that masks are uniform regarding the size across the batch or they can vary?
If can vary maybe we can differently split the work across the threads. Ask each thread to create a mask for only one plan in each volume.
Does it make any sense to have it on the GPU?

Reply:
And how that works with 3D masks?

Reply:
I can benchmark it. I wouldn't assume uniform shape.
I don't understand this:
> Ask each thread to create a mask for only one plan in each volume.
SearchableRLEMask works on flat indices. We are not doing anything with the dimensions.

Regarding the GPU, encoding the mask is a rather serial task. Having a mask per plane would complicate things, so I wouldn't go there unless we see there is a performance bottleneck here. On top of that, Slice operator doesn't take GPU anchor/shape.

Reply:
Ok, let us benchmark first and see if there is anything to fight for.